File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/CourseGradingAgentsLogic.kt
package com.lextures.android.core.lms

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/** Course grading agents list/edit helpers (M13.6). */
object CourseGradingAgentsLogic {
    data class GradableOption(
        val id: String,
        val label: String,
        val kind: String,
    )

    data class AgentDraft(
        val prompt: String = "",
        val includeAssignmentContent: Boolean = false,
        val includeRubric: Boolean = false,
        val status: String = AgentStatus.Draft.name.lowercase(),
        val autoGradeNew: Boolean = false,
        val postPolicy: String = "draft",
        val confidenceFloor: Double? = null,
        val workflowGraph: JsonElement? = null,
    )

    enum class AgentStatus {
        Draft,
        Accepted,
        Archived,
        ;

        val apiValue: String
            get() = name.lowercase()

        val labelRes: Int
            get() = when (this) {
                Draft -> com.lextures.android.R.string.mobile_courseSettings_gradingAgents_status_draft
                Accepted -> com.lextures.android.R.string.mobile_courseSettings_gradingAgents_status_accepted
                Archived -> com.lextures.android.R.string.mobile_courseSettings_gradingAgents_status_archived
            }

        companion object {
            fun fromApi(value: String): AgentStatus =
                entries.firstOrNull { it.apiValue == value } ?: Draft
        }
    }

    enum class ValidationError {
        PromptRequired,
    }

    private const val WORKFLOW_VERSION = 1

    fun cacheKeyAgents(courseCode: String): String = "course:$courseCode:grading-agents"

    fun saveIdempotencyKey(courseCode: String, itemId: String, itemKind: String): String =
        "course-grading-agents:$courseCode:$itemKind:$itemId:save"

    fun deleteIdempotencyKey(courseCode: String, itemId: String, itemKind: String): String =
        "course-grading-agents:$courseCode:$itemKind:$itemId:delete"

    fun graderAgentPath(courseCode: String, itemId: String, itemKind: String): String {
        val collection = if (itemKind == "quiz") "quizzes" else "assignments"
        return "/api/v1/courses/$courseCode/$collection/$itemId/grader-agent"
    }

    fun normalizedItemKind(itemKind: String?): String =
        if (itemKind == "quiz") "quiz" else "assignment"

    fun gradableOptions(
        structure: List<CourseStructureItem>,
        existingItemIds: Set<String>,
    ): List<GradableOption> {
        val byId = structure.associateBy { it.id }
        return structure
            .filter { item ->
                (item.kind == "assignment" || item.kind == "quiz") &&
                    item.archived != true &&
                    item.id !in existingItemIds
            }
            .map { item ->
                var moduleTitle = ""
                var parent = item.parentId?.let { byId[it] }
                val guardIds = mutableSetOf<String>()
                while (parent != null && parent.id !in guardIds) {
                    guardIds += parent.id
                    if (parent.kind == "module") {
                        moduleTitle = parent.title
                        break
                    }
                    parent = parent.parentId?.let { byId[it] }
                }
                val label = if (moduleTitle.isEmpty()) item.title else "$moduleTitle — ${item.title}"
                GradableOption(id = item.id, label = label, kind = item.kind)
            }
            .sortedBy { it.label.lowercase() }
    }

    fun draft(config: GraderAgentConfig?): AgentDraft =
        AgentDraft(
            prompt = config?.prompt.orEmpty(),
            includeAssignmentContent = config?.includeAssignmentContent == true,
            includeRubric = config?.includeRubric == true,
            status = config?.status ?: AgentStatus.Draft.apiValue,
            autoGradeNew = config?.autoGradeNew == true,
            postPolicy = config?.postPolicy ?: "draft",
            confidenceFloor = config?.confidenceFloor,
            workflowGraph = config?.workflowGraph,
        )

    fun draft(template: GraderAgentTemplateDetail): AgentDraft =
        AgentDraft(
            prompt = template.prompt,
            includeAssignmentContent = template.includeAssignmentContent,
            includeRubric = template.includeRubric,
            status = AgentStatus.Draft.apiValue,
            autoGradeNew = false,
            postPolicy = "draft",
            confidenceFloor = null,
            workflowGraph = template.workflowGraph,
        )

    fun isDirty(current: AgentDraft, baseline: AgentDraft): Boolean = current != baseline

    fun validateDraft(draft: AgentDraft): ValidationError? =
        if (draft.prompt.trim().isEmpty()) ValidationError.PromptRequired else null

    fun buildPutBody(current: AgentDraft, itemKind: String): PutGraderAgentConfigBody {
        val floor = current.confidenceFloor
        return PutGraderAgentConfigBody(
            prompt = current.prompt.trim(),
            includeAssignmentContent = current.includeAssignmentContent,
            includeRubric = current.includeRubric,
            status = current.status,
            autoGradeNew = current.autoGradeNew,
            postPolicy = current.postPolicy,
            confidenceFloor = floor?.takeIf { it > 0 },
            workflowGraph = effectiveWorkflowGraph(current.workflowGraph, itemKind),
        )
    }

    fun effectiveWorkflowGraph(stored: JsonElement?, itemKind: String): JsonElement {
        if (stored != null && hasWorkflowNodes(stored)) {
            return if (itemKind == "quiz") ensureQuizResponsesNode(stored) else stored
        }
        return defaultWorkflowGraph(itemKind)
    }

    fun filteredAgents(
        agents: List<CourseGradingAgentSummary>,
        query: String,
    ): List<CourseGradingAgentSummary> {
        val trimmed = query.trim()
        val sorted = agents.sortedBy { it.assignmentTitle.lowercase() }
        if (trimmed.isEmpty()) return sorted
        val needle = trimmed.lowercase()
        return sorted.filter { agent ->
            agent.assignmentTitle.lowercase().contains(needle) ||
                agent.status.lowercase().contains(needle)
        }
    }

    fun statusLabelRes(status: String): Int =
        AgentStatus.fromApi(status).labelRes

    fun kindLabelRes(itemKind: String?): Int? =
        if (normalizedItemKind(itemKind) == "quiz") {
            com.lextures.android.R.string.mobile_courseSettings_gradingAgents_quizBadge
        } else {
            null
        }

    private fun hasWorkflowNodes(graph: JsonElement): Boolean {
        val nodes = graph.jsonObject["nodes"]?.jsonArray ?: return false
        return nodes.isNotEmpty()
    }

    private fun ensureQuizResponsesNode(graph: JsonElement): JsonElement {
        val objectGraph = graph.jsonObject
        val nodes = objectGraph["nodes"]?.jsonArray ?: return defaultWorkflowGraph("quiz")
        val hasQuizResponses = nodes.any { node ->
            node.jsonObject["type"]?.jsonPrimitive?.content == "quizResponses"
        }
        if (hasQuizResponses) return graph
        val quizNode = buildJsonObject {
            put("id", JsonPrimitive("quizResponses"))
            put("type", JsonPrimitive("quizResponses"))
            put(
                "position",
                buildJsonObject {
                    put("x", JsonPrimitive(-420))
                    put("y", JsonPrimitive(0))
                },
            )
            put("data", JsonObject(emptyMap()))
        }
        return buildJsonObject {
            objectGraph.forEach { (key, value) ->
                if (key == "nodes") {
                    put(key, JsonArray(listOf(quizNode) + nodes))
                } else {
                    put(key, value)
                }
            }
        }
    }

    fun defaultWorkflowGraph(itemKind: String): JsonElement {
        val outputNode = buildJsonObject {
            put("id", JsonPrimitive("output"))
            put("type", JsonPrimitive("output"))
            put(
                "position",
                buildJsonObject {
                    put("x", JsonPrimitive(0))
                    put("y", JsonPrimitive(0))
                },
            )
            put("data", JsonObject(emptyMap()))
        }
        val nodes = if (itemKind == "quiz") {
            val quizNode = buildJsonObject {
                put("id", JsonPrimitive("quizResponses"))
                put("type", JsonPrimitive("quizResponses"))
                put(
                    "position",
                    buildJsonObject {
                        put("x", JsonPrimitive(-420))
                        put("y", JsonPrimitive(0))
                    },
                )
                put("data", JsonObject(emptyMap()))
            }
            buildJsonArray {
                add(quizNode)
                add(outputNode)
            }
        } else {
            buildJsonArray { add(outputNode) }
        }
        return buildJsonObject {
            put("version", JsonPrimitive(WORKFLOW_VERSION))
            put("nodes", nodes)
            put("edges", JsonArray(emptyList()))
        }
    }
}

File: clients/ios/Lextures/Core/LMS/CourseGradingAgentsLogic.swift
import Foundation

/// Course grading agents list/edit helpers (M13.6).
enum CourseGradingAgentsLogic {
    struct GradableOption: Identifiable, Hashable {
        var id: String
        var label: String
        var kind: String
    }

    struct AgentDraft: Equatable, Hashable {
        var prompt: String
        var includeAssignmentContent: Bool
        var includeRubric: Bool
        var status: String
        var autoGradeNew: Bool
        var postPolicy: String
        var confidenceFloor: Double?
        var workflowGraph: JSONValue?
    }

    enum AgentStatus: String, CaseIterable, Identifiable {
        case draft
        case accepted
        case archived

        var id: String { rawValue }

        var labelKey: String {
            "mobile.courseSettings.gradingAgents.status.\(rawValue)"
        }
    }

    enum ValidationError: Equatable {
        case promptRequired
    }

    private static let workflowVersion = 1

    static func cacheKeyAgents(courseCode: String) -> String {
        "course:\(courseCode):grading-agents"
    }

    static func saveIdempotencyKey(courseCode: String, itemId: String, itemKind: String) -> String {
        "course-grading-agents:\(courseCode):\(itemKind):\(itemId):save"
    }

    static func deleteIdempotencyKey(courseCode: String, itemId: String, itemKind: String) -> String {
        "course-grading-agents:\(courseCode):\(itemKind):\(itemId):delete"
    }

    static func graderAgentPath(courseCode: String, itemId: String, itemKind: String) -> String {
        let collection = itemKind == "quiz" ? "quizzes" : "assignments"
        return "/api/v1/courses/\(courseCode)/\(collection)/\(itemId)/grader-agent"
    }

    static func normalizedItemKind(_ itemKind: String?) -> String {
        itemKind == "quiz" ? "quiz" : "assignment"
    }

    static func gradableOptions(
        from structure: [CourseStructureItem],
        excluding existingItemIds: Set<String>
    ) -> [GradableOption] {
        let byId = Dictionary(uniqueKeysWithValues: structure.map { ($0.id, $0) })
        let rows = structure.filter { item in
            (item.kind == "assignment" || item.kind == "quiz")
                && item.archived != true
                && !existingItemIds.contains(item.id)
        }
        let withLabels: [GradableOption] = rows.map { item in
            var moduleTitle = ""
            var parent = item.parentId.flatMap { byId[$0] }
            var guardIds = Set<String>()
            while let currentParent = parent, !guardIds.contains(currentParent.id) {
                guardIds.insert(currentParent.id)
                if currentParent.kind == "module" {
                    moduleTitle = currentParent.title
                    break
                }
                parent = currentParent.parentId.flatMap { byId[$0] }
            }
            let label = moduleTitle.isEmpty ? item.title : "\(moduleTitle) — \(item.title)"
            return GradableOption(id: item.id, label: label, kind: item.kind)
        }
        return withLabels.sorted { $0.label.localizedCaseInsensitiveCompare($1.label) == .orderedAscending }
    }

    static func draft(from config: GraderAgentConfig?) -> AgentDraft {
        AgentDraft(
            prompt: config?.prompt ?? "",
            includeAssignmentContent: config?.includeAssignmentContent ?? false,
            includeRubric: config?.includeRubric ?? false,
            status: config?.status ?? AgentStatus.draft.rawValue,
            autoGradeNew: config?.autoGradeNew ?? false,
            postPolicy: config?.postPolicy ?? "draft",
            confidenceFloor: config?.confidenceFloor,
            workflowGraph: config?.workflowGraph
        )
    }

    static func draft(from template: GraderAgentTemplateDetail) -> AgentDraft {
        AgentDraft(
            prompt: template.prompt,
            includeAssignmentContent: template.includeAssignmentContent,
            includeRubric: template.includeRubric,
            status: AgentStatus.draft.rawValue,
            autoGradeNew: false,
            postPolicy: "draft",
            confidenceFloor: nil,
            workflowGraph: template.workflowGraph
        )
    }

    static func isDirty(current: AgentDraft, baseline: AgentDraft) -> Bool {
        current != baseline
    }

    static func validateDraft(_ draft: AgentDraft) -> ValidationError? {
        draft.prompt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? .promptRequired : nil
    }

    static func buildPutBody(current: AgentDraft, itemKind: String) -> PutGraderAgentConfigBody {
        let floor = current.confidenceFloor
        return PutGraderAgentConfigBody(
            prompt: current.prompt.trimmingCharacters(in: .whitespacesAndNewlines),
            includeAssignmentContent: current.includeAssignmentContent,
            includeRubric: current.includeRubric,
            status: current.status,
            autoGradeNew: current.autoGradeNew,
            postPolicy: current.postPolicy,
            confidenceFloor: floor.flatMap { $0 > 0 ? $0 : nil },
            workflowGraph: effectiveWorkflowGraph(
                stored: current.workflowGraph,
                itemKind: itemKind
            )
        )
    }

    static func effectiveWorkflowGraph(stored: JSONValue?, itemKind: String) -> JSONValue {
        if let stored, hasWorkflowNodes(stored) {
            return itemKind == "quiz" ? ensureQuizResponsesNode(stored) : stored
        }
        return defaultWorkflowGraph(itemKind: itemKind)
    }

    static func filteredAgents(
        _ agents: [CourseGradingAgentSummary],
        query: String
    ) -> [CourseGradingAgentSummary] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return agents.sorted { $0.assignmentTitle.localizedCaseInsensitiveCompare($1.assignmentTitle) == .orderedAscending }
        }
        let needle = trimmed.lowercased()
        return agents
            .filter { agent in
                agent.assignmentTitle.lowercased().contains(needle)
                    || statusLabelKey(agent.status).lowercased().contains(needle)
            }
            .sorted { $0.assignmentTitle.localizedCaseInsensitiveCompare($1.assignmentTitle) == .orderedAscending }
    }

    static func statusLabelKey(_ status: String) -> String {
        AgentStatus(rawValue: status)?.labelKey ?? status
    }

    static func kindLabelKey(_ itemKind: String?) -> String? {
        normalizedItemKind(itemKind) == "quiz"
            ? "mobile.courseSettings.gradingAgents.quizBadge"
            : nil
    }

    private static func hasWorkflowNodes(_ graph: JSONValue) -> Bool {
        guard case .object(let object) = graph,
              case .array(let nodes)? = object["nodes"] else {
            return false
        }
        return !nodes.isEmpty
    }

    private static func ensureQuizResponsesNode(_ graph: JSONValue) -> JSONValue {
        guard case .object(var object) = graph,
              case .array(let nodes) = object["nodes"] else {
            return defaultWorkflowGraph(itemKind: "quiz")
        }
        let hasQuizResponses = nodes.contains { node in
            guard case .object(let nodeObject) = node,
                  case .string(let type)? = nodeObject["type"] else {
                return false
            }
            return type == "quizResponses"
        }
        guard !hasQuizResponses else { return graph }
        let quizNode: JSONValue = .object([
            "id": .string("quizResponses"),
            "type": .string("quizResponses"),
            "position": .object(["x": .number(-420), "y": .number(0)]),
            "data": .object([:]),
        ])
        object["nodes"] = .array([quizNode] + nodes)
        return .object(object)
    }

    static func defaultWorkflowGraph(itemKind: String) -> JSONValue {
        let outputNode: JSONValue = .object([
            "id": .string("output"),
            "type": .string("output"),
            "position": .object(["x": .number(0), "y": .number(0)]),
            "data": .object([:]),
        ])
        if itemKind == "quiz" {
            let quizNode: JSONValue = .object([
                "id": .string("quizResponses"),
                "type": .string("quizResponses"),
                "position": .object(["x": .number(-420), "y": .number(0)]),
                "data": .object([:]),
            ])
            return .object([
                "version": .number(Double(workflowVersion)),
                "nodes": .array([quizNode, outputNode]),
                "edges": .array([]),
            ])
        }
        return .object([
            "version": .number(Double(workflowVersion)),
            "nodes": .array([outputNode]),
            "edges": .array([]),
        ])
    }
}
